core/src/main/java/org/infinispan/util/concurrent/locks/impl/InfinispanLock.java

```diff
@@ -0,0 +1,555 @@
+      if (trace) {
+         log.tracef("Release lock for %s. It was canceled.", canceled.lockOwner);
+      }
+      LockPlaceHolder wantToRelease = lockOwners.get(canceled.lockOwner);
+      if (wantToRelease == null) {
+         if (trace) {
+            log.tracef("%s not found!", canceled.lockOwner);
+         }
+         //nothing to release
+         return;
+      }
+      LockPlaceHolder currentLocked = current;
+      if (currentLocked == wantToRelease) {
+         if (!casRelease(currentLocked)) {
+            if (trace) {
+               log.tracef("Releasing lock for %s. It is the current lock owner but another thread changed it.", canceled.lockOwner);
+            }
+            //another thread already released!
+         } else {
+            tryAcquire();
+         }
+      }
+   }
+
+   private boolean casAcquire(LockPlaceHolder lockPlaceHolder) {
+      return fieldUpdater.compareAndSet(this, null, lockPlaceHolder);
+   }
+
+   private boolean casRelease(LockPlaceHolder lockPlaceHolder) {
+      return fieldUpdater.compareAndSet(this, lockPlaceHolder, null);
+   }
+
+   private void remove(Object lockOwner) {
+      lockOwners.remove(lockOwner);
+   }
+
+   private void triggerReleased() {
+      if (releaseRunnable != null) {
+         releaseRunnable.run();
+      }
+   }
+
+   private void tryAcquire() {
+      do {
+         LockPlaceHolder nextPending = pendingRequest.peek();
+         if (trace) {
+            log.tracef("Try acquire. Next in queue=%s. Current=%s", nextPending, current);
+         }
+         if (nextPending == null) {
+            return;
+         }
+         if (casAcquire(nextPending)) {
+            //we set the current lock owner, so we must remove it from the queue
+            pendingRequest.remove(nextPending);
+            if (nextPending.setAvailable()) {
+               if (trace) {
+                  log.tracef("%s successfully acquired the lock.", nextPending);
+               }
+               //successfully acquired
+               deadlockCheck();
+               return;
+            }
+            if (trace) {
+               log.tracef("%s failed to acquire (invalid state). Retrying.", nextPending);
+            }
+            //oh oh, probably the nextPending Timed-Out. we are going to retry with the next in queue
+            casRelease(nextPending);
+         } else {
+            if (trace) {
+               log.tracef("Unable to acquire. Lock is held.");
+            }
+            //other thread already set the current lock owner
+            return;
+         }
+      } while (true);
+   }
+
+   private LockPlaceHolder createLockInfo(Object lockOwner, long time, TimeUnit timeUnit) {
+      return new LockPlaceHolder(lockOwner, timeService.expectedEndTime(time, timeUnit));
+   }
+
+   private class LockPlaceHolder implements ExtendedLockPromise {
+
+      private final AtomicReferenceFieldUpdater<LockPlaceHolder, LockState> stateUpdater;
+
+      private final Object lockOwner;
```

Comment:
I'm not sure I'd call it `lockOwner`, because it doesn't own the lock yet, it only owns this placeholder...